docs/imagelayout.1
.\" Automatically generated by Pandoc 2.9.2.1
.\"
.TH "imagelayout.py" "1" "July 2019" "imagelayout.py v1.0" "Version 1.0"
.hy
.SH NAME
.PP
\f[B]imagelayout\f[R] - arrange images into a large image according to a
layout
.SH SYNOPSYS
.PP
\f[B]imagelayout(.py)\f[R] [\f[B]-w\f[R]|\f[B]--watch\f[R]]
[\f[B]-s\f[R]] [\f[B]-o\f[R] \f[I]outputfile\f[R]] [\f[B]-h|--help\f[R]]
\f[I]configfile\f[R] [\f[I]imagefile\f[R] [\f[I]imagefile\f[R] ...]]
.PP
NOTE: Depending on how you installed the program, the command is either
\f[B]imagelayout.py\f[R] or just \f[B]imagelayout\f[R].
.SH DESCRIPTION
.PP
\f[B]imagelayout\f[R] arranges several images according to a predefined
layout and produces a larger image consisting of the sub-images.
It can optionally label the individual images.
Additional features include the ability to crop and add borders to the
images, to add a border and a title to the final image, and to add
arbitrary labels and lines or arrows to the final image.
The intended use of \f[B]imagelayout\f[R] is the creation of figures
from sub-figures for articles and science publications.
\f[B]imagelayout\f[R] is a command-line Python application guided by a
configuration file \f[I]configfile\f[R] (provided as the only command
line argument); it has no graphical interface or interactive features.
Thus, it is suitable for use in scripts, Makefiles, and automated and
reproducible workflows.
.SH OPTIONS
.IP \[bu] 2
\f[B]-w, --watch\f[R]
.PD 0
.P
.PD
Run once, then watch the configuration file for changes and re-run
whenever a
.PD 0
.P
.PD
change is detected.
This is useful for developing the configuration file and making
adjustments.
\f[B]Tip:\f[R] View the output image file with a viewer that also
watches the image file and reloads it when it changes.
\f[C]qiv\f[R] for Linux systems is such a viewer, e.g.
\f[C]qiv -T out.png\f[R] will reload \f[C]out.png\f[R] as soon as it
changes.
.IP \[bu] 2
\f[B]-s\f[R]
.PD 0
.P
.PD
Report the pixel sizes of input images and exit.
.IP \[bu] 2
\f[B]-o\f[R] \f[I]outputfile\f[R]
.PD 0
.P
.PD
Name of the output image file (optional)
.IP \[bu] 2
\f[B]-h, --help\f[R]
.PD 0
.P
.PD
Display short help message and exit.
.IP \[bu] 2
\f[I]configfile\f[R]
.PD 0
.P
.PD
The configuration file (yaml format)
.IP \[bu] 2
\f[I]imagefile(s)\f[R]
.PD 0
.P
.PD
Optionally, the image files to be used as inputs can be specified here.
.SH DEFINING THE IMAGE LAYOUT
.PP
The desired layout of the images can be defined as a combination of
horizontally and vertically joined images.
Two keywords are used for this in the configuration file.
The \f[C]hjoin\f[R] keyword defines a horizontally joined image
sequence, e.g.
.IP
.nf
\f[C]
layout:
  hjoin: [A, B, C]
\f[R]
.fi
.PP
will result in the following image arrangement:
.IP
.nf
\f[C]
+---+---+---+
| A | B | C |
+---+---+---+
\f[R]
.fi
.PP
while the \f[C]vjoin\f[R] keyword produces a vertical arrangement:
.IP
.nf
\f[C]
layout:
  vjoin: [A, B, C]
\f[R]
.fi
.PP
will result in this arrangement:
.IP
.nf
\f[C]
+---+
| A |
+---+
| B |
+---+
| C |
+---+
\f[R]
.fi
.PP
More complex layouts can be defined by combining and nesting the
\f[C]hjoin\f[R] and \f[C]vjoin\f[R] specifications.
For example,
.IP
.nf
\f[C]
hjoin:
  - vjoin:
      - hjoin: [A, B]
      - C
  - D
\f[R]
.fi
.PP
results in this layout:
.IP
.nf
\f[C]
+---+---+---+
| A | B |   |
+---+---+ D |
|   C   |   |
+-------+---+
\f[R]
.fi
.PP
Horizontally joined images will be resized to fit the same height (the
highest image), and vertically joined images will be resized to fit the
same width (the widest image).
If the \f[B]fixedsize\f[R] option is used for an image, it will be
padded rather than scaled.
.SH INPUT AND OUTPUT FILE NAMES
.PP
The names of the input and output image file names can be specified in
the configuration file.
However, to facilitate the use of \f[C]imagelayout.py\f[R] in scripts
and Makefiles, the file names can also be provided on the command line.
.PP
Input file names can be provided on the command line after the
configuration file name.
In this case, you can refer to them in the configuration file as
\f[C]$1\f[R], \f[C]$2\f[R], etc.
For example, if the command line is
.IP
.nf
\f[C]
imagelayout.py config.yaml image1.png image2.jpg image3.png
\f[R]
.fi
.PP
then you can refer to the three input image files in the configuration
file as follows:
.IP
.nf
\f[C]
images:
  A: $1
  B: $2
  C: $3
\f[R]
.fi
.PP
To specify the output file name on the command line, the \f[C]-o\f[R]
option can be used.
In this case, you can refer to it in the configuration file as any
string starting with a \f[C]$\f[R] (a single \f[C]$\f[R] is enough).
For example, if the command line is
.IP
.nf
\f[C]
imagelayout.py -o output.png config.yaml
\f[R]
.fi
.PP
then you refer to the output file in the configuration file as follows:
.IP
.nf
\f[C]
outputfile: $
\f[R]
.fi
.PP
You can use special file names of the format
\f[C]BLANK-\f[R]W\f[C]x\f[R]H, e.g.
\f[C]BLANK-640x480\f[R].
In this case, a blank image will be generated by
\f[C]imagelayout.py\f[R] with the given width and height.
The color of the blank image will be determined by the
\f[C]paddingcolor\f[R] keyword (or white by default).
The blank images can be used as placeholders.
For example, if you have 5 images and you want to arrange them into a
3x2 grid and leave the bottom right cell empty then you can use the
following configuration:
.IP
.nf
\f[C]
images:
  A: A.png
  B: B.png
  C: C.png
  D: D.png
  E: E.png
  blank: BLANK-640x480
layout:
  vjoin:
    - hjoin: [A, B, C]
    - hjoin: [D, E, blank]
\f[R]
.fi
.PP
The resulting image will look like this:
.IP
.nf
\f[C]
+---+---+---+
| A | B | C |
+---+---+---+
| D | E |   |
+---+---+---+
\f[R]
.fi
.PP
By default, the identifiers of the blank images will not be displayed in
the final image.
.SH CONFIGURATION FILE
.PP
The only mandatory command line argument to \f[B]imagelayout\f[R] is the
configuration file \f[I]configfile\f[R], which should be written as a
YAML format file, and contains all information about the input/output
files, the layout, and the requested operations.
.PP
The two main mandatory parts of the configuration file are the
\f[B]images\f[R] section and the \f[B]layout\f[R] section; in addition
the output file must also be specified using the \f[B]outputfile\f[R]
keyword.
The simplest configuration file may look something like this:
.IP
.nf
\f[C]
images:
  A: imageA.png
  B: imageB.png
outputfile: out.png
layout:
  hjoin: [A, B]
\f[R]
.fi
.PP
This will produce the file \f[C]out.png\f[R] consisting of
\f[C]imageA.png\f[R] and \f[C]imageB.png\f[R] joined horizontally.
If the
.IP
.nf
\f[C]
labels:
  add: yes
\f[R]
.fi
.PP
lines are added then each image will be labeled in the top left corner
with the letters A and B, respectively.
.PP
The configuration file may contain numerous other keywords and options
to specify image labels, font names, font sizes and colors, cropping
images or adding borders to them, adding a title, adding arbitrary
labels, lines, and arrows to the final image.
The detailed list of keywords and options follows.
.PP
\f[B]Note on font names:\f[R] On Linux, the system fonts may be listed
using the \f[C]fc-list\f[R] command.
On Windows, look in the \f[C]\[rs]Windows\[rs]Fonts\f[R] directory.
Use the font file names without the extension (e.g.
\f[C].ttf\f[R]) as font names.
For example, bold Arial is specified in Linux as \f[C]Arial_Bold\f[R]
(assuming the \f[C]msttcorefonts\f[R] package is installed), but on
Windows as \f[C]arialbd\f[R] (as the \f[C]\[rs]Windows\[rs]Fonts\f[R]
directory contains the \f[C]arialbd.ttf\f[R] file).
.SH COMPLETE LIST OF CONFIGURATION FILE OPTIONS
.SS imagelayout yaml configuration file
.IP \[bu] 2
\f[B]inputdir:\f[R] \f[I]\f[BI]directoryname\f[I]\f[R]
.PD 0
.P
.PD
Input directory; input image files will be loaded from here
.IP \[bu] 2
\f[B]images:\f[R]
.PD 0
.P
.PD
Input images are specified here.
Each file must have an identifier, and either a single file name or a
number of parameters.
Note: A file name like \f[C]$1\f[R], \f[C]$2\f[R], etc.
indicates that the file name should be taken from the command line (see
manual).
If the file name is \f[C]BLANK-\f[R]W\f[C]x\f[R]H, e.g.
\f[C]BLANK-640x480\f[R] then a blank image of the given size will be
generated; this can be used as a placeholder; identifiers of blank
images will not be displayed as labels.
.RS 2
.IP \[bu] 2
\f[B]\f[BI]imageid\f[B]:\f[R] \f[B]\f[BI]filename\f[B]\f[R]
.PD 0
.P
.PD
An identifier and a filename.
If you don\[aq]t want to specify any further parameters for the image
other than the filename then this form can be used.
Otherwise, see below.
.IP \[bu] 2
\f[B]\f[BI]imageid\f[B]:\f[R]
.PD 0
.P
.PD
An identifier for an image
.RS 2
.IP \[bu] 2
\f[B]file:\f[R] \f[B]\f[BI]filename\f[B]\f[R]
.PD 0
.P
.PD
Name of the image file
.IP \[bu] 2
\f[B]label:\f[R] \f[B]\f[BI]label\f[B]\f[R]
.PD 0
.P
.PD
A label for the image.
A text string if you don\[aq]t want to specify any parameters such as
position, font, color, etc; in this case the parameters specified in the
toplevel \f[B]labels\f[R] option will be used.
If you want to specify further parameters here then provide a mapping
(see next line).
Optional; if omitted then the image identifier will be used for
labeling.
.IP \[bu] 2
\f[B]label:\f[R]
.PD 0
.P
.PD
A label for the image along with parameters such as position, font, etc.
These parameters will override those given in the toplevel
\f[B]labels\f[R] option.
.RS 2
.IP \[bu] 2
\f[B]text:\f[R] \f[B]\f[BI]labeltext\f[B]\f[R]
.PD 0
.P
.PD
Text of the label; if omitted then the image identifier will be used.
.IP \[bu] 2
\f[B]fontname:\f[R] \f[B]\f[BI]fontname\f[B]\f[R]
.PD 0
.P
.PD
Name of a truetype font available on the system.
Example: \f[C]Arial_Bold\f[R].
.IP \[bu] 2
\f[B]fontsize:\f[R] \f[B]\f[BI]size\f[B]\f[R]
.PD 0
.P
.PD
Font size in pixels.
.IP \[bu] 2
\f[B]fontcolor:\f[R] \f[B]\f[BI]colorspec\f[B]\f[R]
.PD 0
.P
.PD
Font color.
A color name like \f[C]red\f[R], an RGB color specified as
\f[C]#rrggbb\f[R] or \f[C]rgb(red, green, blue)\f[R] with values between
0 and 255 or percentages, or a HSL color as
\f[C]hsl(hue, saturation%, lightness%)\f[R].
.IP \[bu] 2
\f[B]pos:\f[R]
(\f[B]\f[CB]top\f[B]\f[R]|\f[C]center\f[R]|\f[C]bottom\f[R])\f[C]-\f[R](\f[B]\f[CB]left\f[B]\f[R]|\f[C]center\f[R]|\f[C]right\f[R])
.PD 0
.P
.PD
Position of the label relative to the image.
A combination of the words (\f[C]top\f[R], \f[C]center\f[R],
\f[C]bottom\f[R]) and (\f[C]left\f[R], \f[C]center\f[R],
\f[C]right\f[R]), separated with a dash (no space in between!).
Default: \f[C]top-left\f[R].
.IP \[bu] 2
\f[B]offset:\f[R] \f[B][ \f[BI]xoffset\f[B]\f[R],
\f[B]\f[BI]yoffset\f[B] ]\f[R]
.PD 0
.P
.PD
Offset, in pixels, for the label, relative to the position defined in
\f[B]pos\f[R].
Default: \f[C][0, 0]\f[R]
.RE
.IP \[bu] 2
\f[B]fixedsize:\f[R] \f[B]\f[CB]no\f[B]\f[R] | \f[C]yes\f[R]
.PD 0
.P
.PD
If \f[C]yes\f[R], the image will never be resized during the layout (it
may still be resized along with the final image if the
\f[B]finalwidth\f[R] or \f[B]finalheight\f[R] option is used).
Instead, padding will be used (either on the top and bottom or on left
and right of the image) to fit the image to its neighbors.
The color of the padding is determined by the \f[B]paddingcolor\f[R]
toplevel parameter (default is white).
.IP \[bu] 2
\f[B]autocrop:\f[R] \f[B]\f[CB]no\f[B]\f[R]|\f[C]yes\f[R]
.PD 0
.P
.PD
If \f[C]yes\f[R], the image will be autocropped using the color of the
top-left pixel.
.IP \[bu] 2
\f[B]crop:\f[R] \f[B]\f[BI]width\f[B] | [ \f[BI]xwidth, ywidth\f[B] ] |
[ \f[BI]leftwidth\f[B], \f[BI]topwidth\f[B], \f[BI]rightwidth\f[B],
\f[BI]bottomwidth\f[B] ]\f[R]
.PD 0
.P
.PD
Crop the image on all four sides.
A single number specifies the width of a uniform crop.
Two numbers refer to the left/right and top/bottom crop widths.
If four numbers are given, the crop widths of all four sides can be
specified separately.
Optional.
.IP \[bu] 2
\f[B]border:\f[R]
.PD 0
.P
.PD
Optionally add a border to the image.
.RS 2
.IP \[bu] 2
\f[B]size:\f[R] \f[B]\f[BI]width\f[B] | [ \f[BI]xwidth\f[B],
\f[BI]ywidth\f[B] ] | [ \f[BI]leftwidth\f[B], \f[BI]topwidth\f[B],
\f[BI]rightwidth\f[B], \f[BI]bottomwidth\f[B] ]\f[R]
.PD 0
.P
.PD
Border size.
A single number, or a list of 2 or 4 numbers.
A single number specifies the width of a uniform border.
Two numbers refer to the left/right and top/bottom widths of the border.
If four numbers are given, the widths of the border on all four sides
can be specified separately.
If not specified then the \f[B]size\f[R] defined under the top-level
\f[B]border\f[R] property will be used; or zero if there is no top-level
\f[B]border\f[R] property.
.IP \[bu] 2
\f[B]color:\f[R] \f[B]\f[BI]colorspec\f[B]\f[R]
.PD 0
.P
.PD
Color of the border.
A color name like \f[C]red\f[R], an RGB color specified as
\f[C]#rrggbb\f[R] or \f[C]rgb(red, green, blue)\f[R] with values between
0 and 255 or percentages, or a HSL color as
\f[C]hsl(hue, saturation%, lightness%)\f[R].
Default: white.
.RE
.RE
.IP \[bu] 2
\f[B]\f[BI]imageid\f[B]:\f[R] ...
.PD 0
.P
.PD
Any number of further images can be specified.
.RE
.IP \[bu] 2
\f[B]layout:\f[R]
.PD 0
.P
.PD
The layout of the output image is specified here.
It can be omitted if there is only one input image.
.RS 2
.IP \[bu] 2
\f[B]vjoin | hjoin:\f[R]
.PD 0
.P
.PD
Join the following images vertically (\f[B]vjoin\f[R]) or horizontally
(\f[B]hjoin\f[R]).
Must be followed by a list of the images to join.
List elements can be further \f[B]hjoin\f[R]/\f[B]vjoin\f[R] lists and
individual images.
The list can be specified either on the same line in bracket notation
(e.g.
\f[C]vjoin: [A, B, C]\f[R]) or on separate lines using the dash
notation, e.g.
.RS 2
.IP
.nf
\f[C]
vjoin:
            - A
            - B
            - C
\f[R]
.fi
.IP \[bu] 2
\f[B]- vjoin: | hjoin: | \f[BI]imageid\f[B]\f[R]
.PD 0
.P
.PD
List element: another list of images to join vertically or horizontally,

imagelayout.py
#!/usr/bin/env python3

'''arrange several images according to a layout'''

import yaml
import sys
import os
import argparse
from time import sleep
from PIL import Image, ImageFont, ImageDraw, ImageChops
from collections import deque, OrderedDict
import math

def autocrop(im):
    '''autocrop image based on color of top left pixel'''
    bg = Image.new(im.mode, im.size, im.getpixel((0,0)))
    diff = ImageChops.difference(im, bg)
    diff = ImageChops.add(diff, diff, 2.0, -100)
    bbox = diff.getbbox()
    if bbox:
        return im.crop(bbox)
    else:
        return im

def addborder(im, size=5, color='white'):
    '''add border to image with given size and color'''
    if type(size) == list:
        if len(size) == 2: # left-right and top-bottom borders are given
            left, top = size
            right, bottom = size
        elif len(size) == 4: # all four borders separately given
            left, top, right, bottom = size
    else: # uniform border size given
        top = left = bottom = right = size
    if top == left == bottom == right == 0:
        return im
    w, h = im.size
    newsize = (w+left+right, h+top+bottom)
    tim = Image.new('RGBA', newsize, color)
    tim.paste(im, (left, top))
    return tim

def getconf(configfile):
    '''load configuration file into dictionary'''
    f = open(configfile)
    conf = yaml.load(f, Loader=yaml.Loader)
    f.close()
    return conf

def imagelayout(conf, reportsizes=False, imagefiles=[], outputfile=''):
    '''generate layout from given configuration'''
    # initialize variables
    
    wd = conf.get('inputdir', os.getcwd())
    
    # set up image list
    
    imgdata = conf['images']
    
    for imgid in imgdata:
        if type(imgdata[imgid]) == str: # only filename is given
            imgdata[imgid] = {'file':imgdata[imgid]}
        if imgdata[imgid]['file'][0] == '$': # file name is taken from command line argument
            imgdata[imgid]['file'] = imagefiles[int(imgdata[imgid]['file'][1:])-1]
        if imgdata[imgid]['file'].startswith('BLANK-'): # use a blank image
            w, h = [int(s) for s in imgdata[imgid]['file'].split('-')[1].split('x')]
            im = Image.new('RGBA', (w, h), color=conf.get('paddingcolor', 'white'))
            if 'label' not in imgdata[imgid]:
                imgdata[imgid]['label'] = {'text': ''}
            elif 'text' not in imgdata[imgid]['label']:
                imgdata[imgid]['label']['text'] = ''
        else:
            im = Image.open(os.path.join(wd, imgdata[imgid]['file']))
        imgdata[imgid]['size'] = (0, 0, *im.size)
        imgdata[imgid]['parts'] = [imgid]
        imgdata[imgid]['img'] = im
    
    if reportsizes: # -s option is given, report sizes and exit
        for imgid in imgdata:
            w, h = imgdata[imgid]['size'][2:]
            print('%s %s: %dx%d aspect=%f' % (imgid, imgdata[imgid]['file'], w, h, w/h))
        sys.exit()
    
    # autocrop, crop images if requested
    
    if conf.get('autocrop', False):
        for imgid in imgdata:
            if not imgdata[imgid].get('autocrop', True):
                continue # do not autocrop this image
            imb = autocrop(imgdata[imgid]['img'])
            imgdata[imgid]['img'] = imb
            imgdata[imgid]['size'] = (0, 0, *imb.size)
    
    for imgid in imgdata:
        if imgdata[imgid].get('autocrop'):
            imb = autocrop(imgdata[imgid]['img'])
            imgdata[imgid]['img'] = imb
            imgdata[imgid]['size'] = (0, 0, *imb.size)
        if 'crop' in imgdata[imgid]: # cropping defined for individual image
            size = imgdata[imgid]['crop']
            if type(size) == list:
                if len(size) == 2:
                    left, top = size
                    right, bottom = size
                elif len(size) == 4:
                    left, top, right, bottom = size
            else:
                top = left = bottom = right = size
            if top == left == bottom == right == 0:
                continue
            w, h = imgdata[imgid]['size'][2:]
            x1, y1 = left, top
            x2, y2 = w-right, h-bottom
            im = imgdata[imgid]['img']
            imb = im.crop((x1, y1, x2, y2))
            imgdata[imgid]['img'] = imb
            imgdata[imgid]['size'] = (0, 0, *imb.size)
            
    # add borders to images if requested
    
    bordersize = 0
    bordercolor = 'white'
    if 'border' in conf:
        bordersize = conf['border'].get('size', 10)
        bordercolor = conf['border'].get('color', 'white')
    
    for imgid in imgdata:
        bs = bordersize
        bc = bordercolor
        if 'border' in imgdata[imgid]: # border defined for individual image
            bs = imgdata[imgid]['border'].get('size', bordersize)
            bc = imgdata[imgid]['border'].get('color', bordercolor)
        im = imgdata[imgid]['img']
        imb = addborder(im, bs, bc)
        imgdata[imgid]['img'] = imb
        imgdata[imgid]['size'] = (0, 0, *imb.size)
            
    # determine final sizes
    
    if 'layout' not in conf:
        if len(imgdata) == 1:
            conf['layout'] = {'hjoin': list(imgdata)}
        else:
            raise ValueError('layout not specified')
    layout = conf['layout']
    
    # topological sorting of directed graph defining the layout
    Q = deque()
    Q.appendleft(layout)
    S = [layout]
    
    c = 1
    
    while Q:
        v = Q.pop()
        if type(v) == dict:
            jointype, joinlist = list(v.items())[0]
            jointypec = jointype+str(c)
            v[jointypec] = v.pop(jointype)
            c += 1
            for w in v[jointypec]:
                if type(w) == dict and w not in S:
                    S.append(w)
                    Q.appendleft(w)
    
    adj = OrderedDict() # stores hjoin and vjoin lists
    for dic in S[::-1]:
        name = list(dic)[0]
        ilist = []
        for e in dic[name]:
            if type(e) == dict:
                ilist.append(list(e)[0])
            else:
                ilist.append(e)
        adj[name] = ilist
    
    # determine positions and sizes in final image
    
    for iid in imgdata:
        if 'fixedsize' not in imgdata[iid]:
            imgdata[iid]['fixedsize'] = False

    # do the joinings
    for name in adj:
        imgdata[name] = {'fixedsize': False}
        tomerge = adj[name]
        if True in [imgdata[iid]['fixedsize'] for iid in tomerge]:
            imgdata[name]['fixedsize'] = True
        widths = [imgdata[iid]['size'][2]-imgdata[iid]['size'][0] for iid in tomerge]
        heights = [imgdata[iid]['size'][3]-imgdata[iid]['size'][1] for iid in tomerge]
        fixs = [imgdata[iid]['fixedsize'] for iid in tomerge]
        # find largest height/width image, others will be aligned to this
        if name.startswith('vjoin'):
            im0 = widths.index(max(widths))
        elif name.startswith('hjoin'):
            im0 = heights.index(max(heights))
        # generate vertical join
        if name.startswith('vjoin'):
            wfac = [widths[im0]/widths[j] for j in range(len(widths))]
            hfac = [1 if fixs[j] else wfac[j] for j in range(len(widths))]
            newheights = [hfac[j]*heights[j] for j in range(len(widths))]
            toth = sum(newheights)
            totw = widths[im0]
            for j in range(len(tomerge)):
                kx, ky, kw, kh = imgdata[tomerge[j]]['size']
                jw = kw if fixs[j] else widths[im0]
                jh = newheights[j]
                jx = widths[im0]/2-kw/2 if fixs[j] else 0
                jy = sum(newheights[:j])
                imgdata[tomerge[j]]['size'] = (jx, jy, jw, jh)
                for part in imgdata[tomerge[j]]['parts']:
                    if part == tomerge[j]:
                        continue
                    kx, ky, kw, kh = imgdata[part]['size']
                    if fixs[j]:
                        imgdata[part]['size'] = (jx+kx, jy+ky, kw, kh)
                    else:
                        imgdata[part]['size'] = (wfac[j]*kx, jy+wfac[j]*ky, wfac[j]*kw, wfac[j]*kh)
        # generate horizontal join
        elif name.startswith('hjoin'):
            hfac = [heights[im0]/heights[j] for j in range(len(heights))]
            wfac = [1 if fixs[j] else hfac[j] for j in range(len(heights))]
            newwidths = [wfac[j]*widths[j] for j in range(len(heights))]
            toth = heights[im0]
            totw = sum(newwidths)
            for j in range(len(tomerge)):
                kx, ky, kw, kh = imgdata[tomerge[j]]['size']
                jw = newwidths[j]
                jh = kh if fixs[j] else heights[im0]
                jx = sum(newwidths[:j])
                jy = heights[im0]/2-kh/2 if fixs[j] else 0
                imgdata[tomerge[j]]['size'] = (jx, jy, jw, jh)
                for part in imgdata[tomerge[j]]['parts']:
                    if part == tomerge[j]:
                        continue
                    kx, ky, kw, kh = imgdata[part]['size']
                    if fixs[j]:
                        imgdata[part]['size'] = (jx+kx, jy+ky, kw, kh)
                    else:
                        imgdata[part]['size'] = (jx+hfac[j]*kx, hfac[j]*ky, hfac[j]*kw, hfac[j]*kh)
        # set size and parts for this join
        imgdata[name]['size'] = (0, 0, totw, toth)
        imgdata[name]['parts'] = sum((imgdata[iid]['parts'] for iid in tomerge), [])
    last = list(adj)[-1]
    totw, toth = imgdata[last]['size'][2:]
    
    # resize whole image if requested
    
    eids = [iid for iid in imgdata if not iid.startswith('hjoin') and not iid.startswith('vjoin')]
    if 'finalwidth' in conf or 'finalheight' in conf:
        if 'finalwidth' in conf:
            facw = conf['finalwidth']/totw
            if 'finalheight' in conf:
                fach = conf['finalheight']/toth
            else:
                fach = facw
        else:
            fach = conf['finalheight']/toth
            facw = fach
        for iid in eids:
            [x, y, w, h] = imgdata[iid]['size']
            imgdata[iid]['size'] = [facw*x, fach*y, facw*w, fach*h]
        totw = facw*totw
        toth = fach*toth
        
    # convert coords and sizes to integers by sum-preserving rounding
    
    for iid in eids:
        [x, y, w, h] = imgdata[iid]['size']
        imgdata[iid]['intsize'] = [round(x), round(y), round(x+w)-round(x), round(y+h)-round(y)]
    totwint = round(totw)
    tothint = round(toth)
    
    # create new image and paste each input image to its place
    
    reszmethod = conf.get('resizemethod', 'nearest')
    resmeth = {'nearest':Image.NEAREST, 'bilinear':Image.BILINEAR, 'bicubic':Image.BICUBIC,
      'lanczos':Image.LANCZOS}[reszmethod]
    
    imnew = Image.new('RGBA', (totwint, tothint), color=conf.get('paddingcolor', 'white'))
    for imgid in eids:
        [x, y, w, h] = imgdata[imgid]['intsize']
        im = imgdata[imgid]['img']
        imr = im.resize((w, h), resample=resmeth)
        imnew.paste(imr, (x, y))
    
    # add labels if requested
    
    if conf.get('labels', {'add': False}).get('add', False):
        fontname = conf['labels'].get('fontname', 'FreeSans')
        fontsize = conf['labels'].get('fontsize', 32)
        fontcolor = conf['labels'].get('fontcolor', 'black')
        defaultpos = conf['labels'].get('pos', 'top-left')
        defaultoffset = conf['labels'].get('offset', [0, 0])
        draw = ImageDraw.Draw(imnew)
        for iid in eids:
            label = imgdata[iid].get('label', iid)
            if type(label) == str:
                labeltext = label
                fn = fontname
                fs = fontsize
                fc = fontcolor
                pos = defaultpos
                offset = defaultoffset
            else:
                labeltext = imgdata[iid]['label'].get('text', iid)
                fn = imgdata[iid]['label'].get('fontname', fontname)
                fs = imgdata[iid]['label'].get('fontsize', fontsize)
                fc = imgdata[iid]['label'].get('fontcolor', fontcolor)
                pos = imgdata[iid]['label'].get('pos', defaultpos)
                offset = imgdata[iid]['label'].get('offset', defaultoffset)
            [x, y, w, h] = imgdata[iid]['intsize']
            posh, posw = pos.split('-')
            offsetx, offsety = offset
            imfont = ImageFont.truetype(font=fn, size=fs)
            if posh+posw != 'topleft':
                (textw, texth) = draw.textsize(labeltext, font=imfont)
                if posh == 'center':
                    y = y+(h-texth)/2
                elif posh == 'bottom':
                    y = y+h-texth
                if posw == 'center':
                    x = x+(w-textw)/2
                elif posw == 'right':
                    x = x+w-textw
            x += offsetx
            y += offsety
            draw.text((x, y), labeltext, fill=fc, font=imfont)
    
    # add title if requested
    
    if 'title' in conf and conf['title'].get('add', True):
        fontname = conf['title'].get('fontname', 'FreeSans')
        fontsize = conf['title'].get('fontsize', 36)
        fontcolor = conf['title'].get('fontcolor', 'black')
        bgcolor = conf['title'].get('bgcolor', 'white')
        height = conf['title'].get('height', int(fontsize*1.1))
        w, h = imnew.size
        imnew2 = Image.new('RGBA', (w, h+height), color=bgcolor)
        imnew2.paste(imnew, (0, height))
        draw = ImageDraw.Draw(imnew2)
        font = ImageFont.truetype(font=fontname, size=fontsize)
        title = conf['title']['text']
        (textw, texth) = draw.textsize(title, font=font)
        x = (w-textw)/2
        y = (height-texth)/2
        draw.text((x, y), title, fill=fontcolor, font=font)
        imnew = imnew2
        
    # add lines, if any
    
    if 'lines' in conf:
        cos30 = math.cos(15/180*math.pi)
        sin30 = math.sin(15/180*math.pi)
        width = conf['lines'].get('width', 3)
        color = conf['lines'].get('color', 'black')
        for linedic in conf['lines']['linelist']:
            coords = linedic['fromto']
            todraw = [(coords[2*i], coords[2*i+1]) for i in range(len(coords)//2)]
            lw = linedic.get('width', width)
            lc = linedic.get('color', color)
            arrowsize = linedic.get('arrowsize', 0)
            draw = ImageDraw.Draw(imnew)
            if arrowsize > 0:
                # draw an arrowhead
                x1, y1, x2, y2 = coords[-4:]
                vx, vy = x1-x2, y1-y2
                v = math.sqrt(vx**2+vy**2)
                vx, vy = arrowsize*vx/v, arrowsize*vy/v
                vxL = cos30*vx-sin30*vy
                vyL = sin30*vx+cos30*vy
                vxR = cos30*vx+sin30*vy
                vyR = -sin30*vx+cos30*vy
                todraw += [(x2+vxL, y2+vyL), (x2, y2), (x2+vxR, y2+vyR), (x2, y2)]
            draw.line(todraw, fill=lc, width=lw, joint='curve')
                
    # add global labels, if any
    
    if 'globallabels' in conf:
        fontname = conf['globallabels'].get('fontname', 'FreeSans')
        fontsize = conf['globallabels'].get('fontsize', 36)
        fontcolor = conf['globallabels'].get('fontcolor', 'black')
        for labdic in conf['globallabels']['labellist']:
            fn = labdic.get('fontname', fontname)
            fs = labdic.get('fontsize', fontsize)
            fc = labdic.get('fontcolor', fontcolor)
            al = labdic.get('align', 'left')
            [x, y] = labdic.get('coords', [0, 0])
            draw = ImageDraw.Draw(imnew)
            font = ImageFont.truetype(font=fn, size=fs)
            draw.text((x, y), labdic['text'], fill=fc, font=font, align=al)
    
    # add global border if requested
    
    if 'globalborder' in conf:
        imnew = addborder(imnew, size=conf['globalborder'].get('size', 20),
          color=conf['globalborder'].get('color', 'white'))
    
    # save result
    if conf['outputfile'][0] == '$':
        conf['outputfile'] = outputfile # from command line argument
    ofile = conf['outputfile'].lower()
    background = Image.new('RGBA', imnew.size, color=conf.get('paddingcolor', 'white'))
    imnew = Image.alpha_composite(background, imnew)
    # pixelscaling as last step before saving
    if 'pixelscaling' in conf:
        if type(conf['pixelscaling']) == list:
            xscale, yscale = conf['pixelscaling']
        else:
            xscale = yscale = conf['pixelscaling']
        h, w = imnew.size
        imnew = imnew.resize((int(xscale*h), int(yscale*w)), resample=resmeth)
    if ofile.endswith('.tif') or ofile.endswith('.tiff'):
        imnew.save(conf['outputfile'], compression='tiff_adobe_deflate')
    else:
        imnew.save(conf['outputfile'])
    print('Output written to %s' % (conf['outputfile']))


## Main program

def main():
    parser = argparse.ArgumentParser(description='Arrange images into a layout',
      epilog='See man page at https://github.com/aszilagyi/imagelayout/docs/manu.md')
    parser.add_argument('-w', '--watch', action='store_true',
      help='Watch the config file and re-run upon detecting a change')
    parser.add_argument('-s', action='store_true', help='report image sizes and exit')
    parser.add_argument('-o', dest='outputfile', help='Output image file name (optional)')
    parser.add_argument('configfile', help='config file')
    parser.add_argument('imagefile', nargs='*', help='image files (optional)')
    a = parser.parse_args()

    imagelayout(getconf(a.configfile), reportsizes=a.s, imagefiles=a.imagefile, 
      outputfile=a.outputfile)
    
    if a.watch:
        print('Watching config file (%s) for changes, press Ctrl-C to quit...' % (a.configfile))
        mtime = os.stat(a.configfile).st_mtime
        while True:
            sleep(0.5)
            mt = os.stat(a.configfile).st_mtime
            if mt != mtime:
                print('Re-running...')
                try:
                    imagelayout(getconf(a.configfile), reportsizes=a.s, imagefiles=a.imagefile,
                      outputfile=a.outputfile)
                except ValueError as e:
                    print('ValueError:', e)
                mtime = mt

if __name__ == '__main__':
    main()
    
